torchaudio/datasets/musdb_hq.py

```diff
@@ -0,0 +1,93 @@
+import os
+from pathlib import Path
+from typing import Tuple, Union, List, Optional
+
+import torch
+import torchaudio
+from torch import Tensor
+from torch.utils.data import Dataset
+from torchaudio import transforms as T
+
+_URL = "https://zenodo.org/record/3338373/files/musdb18hq.zip"
+_EXT = ".wav"
+
+
+class MUSDB_HQ(Dataset):
+    def __init__(
+        self,
+        root: Union[str, Path],
+        sources: List[str],
+        subset: str,
+        samplerate: int = 44100,
+        split: Optional[str] = None,
+        validation: Optional[List[str]] = None,
+    ) -> None:
+        self.sources = sources
+        self.user_samplerate = samplerate
+        self.split = split
+        self.validation = validation
+
+        basename = os.path.basename(_URL)
+        basename = basename.rsplit(".", 2)[0]
+        assert subset in ["test", "train"], "`subset` must be one of ['test', 'train']"
+        assert self.split is None or self.split in [
+            "train",
+            "validation",
+        ], "`split` must be one of ['train', 'validation']"
+        assert (
+            self.split is None and self.validation is None or self.split and self.validation
+        ), "`split` and `validation` must both be None or not be None"
+        self._path = os.path.join(root, basename)
+        self._path = os.path.join(self._path, subset)
+        self.names = self._collect_songs()
+
+    def _get_file(self, name, source):
+        return Path(self._path) / name / f"{source}{_EXT}"
+
+    def _load_sample(self, n: int) -> Tuple[torch.Tensor, int, str]:
+        name = self.names[n]
+        wavs = []
+        sample_rate = None
+        for source in self.sources:
+            file = self._get_file(name, source)
+            wav, new_samplerate = torchaudio.load(str(file))
```

Comment:
Since `sample_rate` is in the argument of initialization, we can name it `sample_rate_file` here and assert it equal to `sellf.sample_rate`.

Reply:
I think the point of this is to check that all sample rates across the sources within a track are the same, which may not be the same as the desired output sample rate (which is what I imagine the `sample_rate` in init args represents). In that case, I don't actually think we should keep `sample_rate` as an input arg (user can perform resampling on their own using transforms), and this function should just return the original wav, sample_rate, and track_name.

Reply:
I see. I prefer leaving the resample function in users' code and return the original sample rate of the dataset. Considering DataLoader fetches data happens in CPU, users may prefer resampling waveforms in collator function which can be in GPU and much faster.
If we make sure all waveforms have the same sample_rate, we can remove the assertion here and return the sample rate directly.